File: packages/provide-uterm-go/hub/clock.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package hub

import (
	"context"
	"sync"
	"time"
)

// Clock abstracts the two time sources the Python services read
// (time.monotonic and time.time) plus a cancellable sleep used by the polling
// coordinator. Injecting a Clock keeps every time-dependent service
// deterministic in tests.
//
// Monotonic returns a strictly non-decreasing seconds value (the analogue of
// Python's time.monotonic); Wall returns Unix wall-clock seconds (time.time).
// Sleep blocks for the given number of seconds or until ctx is cancelled.
type Clock interface {
	Monotonic() float64
	Wall() float64
	Sleep(ctx context.Context, seconds float64) error
}

// realClock is the production [Clock]. Monotonic is measured against a base
// captured at construction so the returned value is a real monotonic reading.
type realClock struct{ base time.Time }

// NewRealClock returns a [Clock] backed by the process wall + monotonic clock.
func NewRealClock() Clock { return &realClock{base: time.Now()} }

func (c *realClock) Monotonic() float64 { return time.Since(c.base).Seconds() }

func (c *realClock) Wall() float64 { return float64(time.Now().UnixNano()) / 1e9 }

func (c *realClock) Sleep(ctx context.Context, seconds float64) error {
	if seconds <= 0 {
		return ctx.Err()
	}
	t := time.NewTimer(time.Duration(seconds * float64(time.Second)))
	defer t.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-t.C:
		return nil
	}
}

// orDefaultClock returns c, or a fresh real clock when c is nil, so callers can
// pass nil to opt into the production clock.
func orDefaultClock(c Clock) Clock {
	if c == nil {
		return NewRealClock()
	}
	return c
}

// ManualClock is a deterministic [Clock] for tests and for wave-B composition
// that wants to drive time explicitly. Monotonic and Wall return the values
// last set; Sleep records its argument and advances Monotonic by Step (default
// 1.0), mirroring the fake clock used by the Python polling suite.
type ManualClock struct {
	mu             sync.Mutex
	mono           float64
	wall           float64
	step           float64
	sleepsRecorded []float64
}

// NewManualClock returns a ManualClock starting at monotonic 0 and the given
// wall time, advancing monotonic by 1.0 per Sleep.
func NewManualClock(wall float64) *ManualClock {
	return &ManualClock{wall: wall, step: 1.0}
}

// SetStep overrides the per-Sleep monotonic advance.
func (c *ManualClock) SetStep(step float64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.step = step
}

// SetMonotonic overrides the current monotonic value.
func (c *ManualClock) SetMonotonic(v float64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.mono = v
}

// SetWall overrides the current wall value.
func (c *ManualClock) SetWall(v float64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.wall = v
}

// Monotonic implements [Clock].
func (c *ManualClock) Monotonic() float64 {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.mono
}

// Wall implements [Clock].
func (c *ManualClock) Wall() float64 {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.wall
}

// Sleep records seconds and advances monotonic by Step; it never blocks.
func (c *ManualClock) Sleep(_ context.Context, seconds float64) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.sleepsRecorded = append(c.sleepsRecorded, seconds)
	c.mono += c.step
	return nil
}

// Sleeps returns a copy of the recorded Sleep arguments in call order.
func (c *ManualClock) Sleeps() []float64 {
	c.mu.Lock()
	defer c.mu.Unlock()
	out := make([]float64, len(c.sleepsRecorded))
	copy(out, c.sleepsRecorded)
	return out
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/Hub/Clock.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

namespace Provide.Uterm.Hub;

/// <summary>
/// Clock abstraction matching Python time.monotonic / time.time plus cancellable sleep.
/// </summary>
public interface IClock
{
    double Monotonic();
    double Wall();
    Task SleepAsync(double seconds, CancellationToken cancellationToken = default);
}

/// <summary>Production clock backed by wall + process-relative monotonic time.</summary>
public sealed class RealClock : IClock
{
    private readonly long _baseTicks = Environment.TickCount64;

    public double Monotonic() => (Environment.TickCount64 - _baseTicks) / 1000.0;

    public double Wall() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;

    public async Task SleepAsync(double seconds, CancellationToken cancellationToken = default)
    {
        if (seconds <= 0)
        {
            cancellationToken.ThrowIfCancellationRequested();
            return;
        }

        await Task.Delay(TimeSpan.FromSeconds(seconds), cancellationToken).ConfigureAwait(false);
    }
}

/// <summary>Deterministic clock for tests — Sleep advances monotonic by <see cref="Step"/>.</summary>
public sealed class ManualClock : IClock
{
    private readonly object _gate = new();
    private double _mono;
    private double _wall;
    private double _step = 1.0;
    private readonly List<double> _sleeps = new();

    public ManualClock(double wall = 0)
    {
        _wall = wall;
    }

    public double Step
    {
        get { lock (_gate) return _step; }
        set { lock (_gate) _step = value; }
    }

    public void SetMonotonic(double v)
    {
        lock (_gate) _mono = v;
    }

    public void SetWall(double v)
    {
        lock (_gate) _wall = v;
    }

    public double Monotonic()
    {
        lock (_gate) return _mono;
    }

    public double Wall()
    {
        lock (_gate) return _wall;
    }

    public Task SleepAsync(double seconds, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            _sleeps.Add(seconds);
            _mono += _step;
        }

        return Task.CompletedTask;
    }

    public IReadOnlyList<double> Sleeps()
    {
        lock (_gate) return _sleeps.ToArray();
    }
}

internal static class ClockUtil
{
    public static IClock OrDefault(IClock? clock) => clock ?? new RealClock();
}
